src/hosting/config.ts

```diff
@@ -0,0 +1,256 @@
+import { bold } from "colorette";
+import { cloneDeep, logLabeledWarning } from "../utils";
+
+import { FirebaseError } from "../error";
+import { HostingMultiple, HostingSingle, FirebaseConfig } from "../firebaseConfig";
+import { Options } from "../options";
+import { partition } from "../functional";
+import { Implements, RequireAtLeastOne } from "../metaprogramming";
+import { dirExistsSync } from "../fsutils";
+import { resolveProjectPath } from "../projectPath";
+import path from "path";
+
+// TODO: Consider putting this type along with the Options type.
+// I haven't tried to do this yet because true options embeds classes, not just
+// interfaces.
+// We should consider either refactoring options so it can be more easily mocked,
+// or creating a utility for constructing options easily in tests.
+export interface MockableOptions {
+  project?: string;
+  site?: string;
+  config: {
+    src: FirebaseConfig;
+  };
+  rc: {
+    requireTarget(project: string, type: string, name: string): string[];
+  };
+  cwd?: string;
+  configPath?: string;
+  only?: string;
+  except?: string;
+  normalizedHostingConfig?: HostingMultiple;
+}
+
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+const mockableOptionsIsCompatibleWithOptions: Implements<Options, MockableOptions> = true;
+
+// assertMatches allows us to throw when an --only flag doesn't match a target
+// but an --except flag doesn't. Is this desirable behavior?
+function matchingConfigs(
+  configs: HostingMultiple,
+  targets: string[],
+  assertMatches: boolean
+): HostingMultiple {
+  const matches: HostingMultiple = [];
+  const [hasSite, hasTarget] = partition(configs, (c) => "site" in c);
+  for (const target of targets) {
+    const siteMatch = hasSite.find((c) => c.site === target);
+    const targetMatch = hasTarget.find((c) => c.target === target);
+    if (siteMatch) {
+      matches.push(siteMatch);
+    } else if (targetMatch) {
+      matches.push(targetMatch);
+    } else if (assertMatches) {
+      throw new FirebaseError(
+        `Hosting site or target ${bold(target)} not detected in firebase.json`
+      );
+    }
+  }
+  return matches;
+}
+
+/**
+ * Returns a subset of configs that match the only string
+ */
+export function filterOnly(configs: HostingMultiple, onlyString?: string): HostingMultiple {
+  if (!onlyString) {
+    return configs;
+  }
+
+  let onlyTargets = onlyString.split(",");
```

Comment:
Dumb question: do we control totally the fields of the Options struct? Are there any situations in which doing a split on "," without stripping whitespace could bite us?